diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 88e2f9d0..2636b809 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -1063,6 +1063,7 @@ mt7996_is_ebf_supported(struct mt7996_phy *phy, struct ieee80211_vif *vif,
 	if (!bfee && sts < 2)
 		return false;
 
+#if 0
 	if (sta->deflink.eht_cap.has_eht) {
 		struct ieee80211_sta_eht_cap *pc = &sta->deflink.eht_cap;
 		struct ieee80211_eht_cap_elem_fixed *pe = &pc->eht_cap_elem;
@@ -1074,6 +1075,7 @@ mt7996_is_ebf_supported(struct mt7996_phy *phy, struct ieee80211_vif *vif,
 			return vif->bss_conf.eht_su_beamformer &&
 			       EHT_PHY(CAP0_SU_BEAMFORMER, pe->phy_cap_info[0]);
 	}
+#endif
 
 	if (sta->deflink.he_cap.has_he) {
 		struct ieee80211_he_cap_elem *pe = &sta->deflink.he_cap.he_cap_elem;
