From 4a7ec1c05e62301d30763d61ab59e6eced4462cf Mon Sep 17 00:00:00 2001
Message-ID: <4a7ec1c05e62301d30763d61ab59e6eced4462cf.1692629957.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Mon, 21 Aug 2023 16:54:07 +0200
Subject: [PATCH] mt76: enable wed by default

Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 mt7915/mmio.c | 2 +-
 mt7996/mmio.c | 3 ++-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/mt7915/mmio.c b/mt7915/mmio.c
index 1621a815..ab82e7b8 100644
--- a/mt7915/mmio.c
+++ b/mt7915/mmio.c
@@ -13,7 +13,7 @@
 #include "../trace.h"
 #include "../dma.h"
 
-static bool wed_enable;
+static bool wed_enable = true;
 module_param(wed_enable, bool, 0644);
 MODULE_PARM_DESC(wed_enable, "Enable Wireless Ethernet Dispatch support");
 
diff --git a/mt7996/mmio.c b/mt7996/mmio.c
index b13cbd5e..7ab9d87d 100644
--- a/mt7996/mmio.c
+++ b/mt7996/mmio.c
@@ -12,8 +12,9 @@
 #include "../trace.h"
 #include "../dma.h"
 
-static bool wed_enable;
+static bool wed_enable = true;
 module_param(wed_enable, bool, 0644);
+MODULE_PARM_DESC(wed_enable, "Enable Wireless Ethernet Dispatch support");
 
 static const struct __base mt7996_reg_base[] = {
 	[WF_AGG_BASE]		= { { 0x820e2000, 0x820f2000, 0x830e2000 } },
-- 
2.41.0

