--- a/arch/arm64/boot/dts/mediatek/mt7622.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7622.dtsi
@@ -832,7 +832,6 @@
 		bus-range = <0x00 0xff>;
 		ranges = <0x82000000 0 0x20000000  0x0 0x20000000  0 0x8000000>;
 		status = "disabled";
-		dma-coherent;
 
 		slot0: pcie@0,0 {
 			reg = <0x0000 0 0 0 0>;
@@ -879,7 +878,6 @@
 		bus-range = <0x00 0xff>;
 		ranges = <0x82000000 0 0x28000000  0x0 0x28000000  0 0x8000000>;
 		status = "disabled";
-		dma-coherent;
 
 		slot1: pcie@1,0 {
 			reg = <0x0800 0 0 0 0>;
