--- a/drivers/net/ethernet/mediatek/mtk_wed.c
+++ b/drivers/net/ethernet/mediatek/mtk_wed.c
@@ -578,15 +578,22 @@ mtk_wed_start(struct mtk_wed_device *dev
 
 static int
 mtk_wed_attach(struct mtk_wed_device *dev)
+	__releases(RCU)
 {
 	struct mtk_wed_hw *hw;
 	int ret = 0;
 
-	if (pci_domain_nr(dev->wlan.pci_dev->bus) > 1)
-		return -ENODEV;
+	RCU_LOCKDEP_WARN(!rcu_read_lock_held(),
+			 "mtk_wed_attach without holding the RCU read lock");
 
-	if (!try_module_get(THIS_MODULE))
-		return -ENODEV;
+	if (pci_domain_nr(dev->wlan.pci_dev->bus) > 1 ||
+	    !try_module_get(THIS_MODULE))
+		ret = -ENODEV;
+
+	rcu_read_unlock();
+
+	if (ret)
+		return ret;
 
 	mutex_lock(&hw_lock);
 
--- a/include/linux/soc/mediatek/mtk_wed.h
+++ b/include/linux/soc/mediatek/mtk_wed.h
@@ -86,7 +86,8 @@ mtk_wed_device_attach(struct mtk_wed_dev
 	dev->ops = rcu_dereference(mtk_soc_wed_ops);
 	if (dev->ops)
 		ret = dev->ops->attach(dev);
-	rcu_read_unlock();
+	else
+		rcu_read_unlock();
 
 	if (ret)
 		dev->ops = NULL;
