diff -Naur a/drivers/net/ethernet/intel/e1000e/Makefile b/drivers/net/ethernet/intel/e1000e/Makefile
--- a/drivers/net/ethernet/intel/e1000e/Makefile	2023-12-15 16:31:15.731964000 +0800
+++ b/drivers/net/ethernet/intel/e1000e/Makefile	2023-03-23 10:12:39.404271000 +0800
@@ -5,7 +5,7 @@
 # Makefile for the Intel(R) PRO/1000 ethernet driver
 #
 
-obj-$(CONFIG_E1000E) += e1000e.o
+obj-m += e1000e.o
 
 e1000e-objs := 82571.o ich8lan.o 80003es2lan.o \
 	       mac.o manage.o nvm.o phy.o \
diff -Naur a/drivers/net/ethernet/intel/Makefile b/drivers/net/ethernet/intel/Makefile
--- a/drivers/net/ethernet/intel/Makefile	2023-12-15 16:31:19.071898000 +0800
+++ b/drivers/net/ethernet/intel/Makefile	2023-03-23 09:32:00.079504000 +0800
@@ -5,7 +5,7 @@
 
 obj-$(CONFIG_E100) += e100.o
 obj-$(CONFIG_E1000) += e1000/
-obj-$(CONFIG_E1000E) += e1000e/
+obj-m += e1000e/
 obj-$(CONFIG_IGB) += igb/
 obj-$(CONFIG_IGC) += igc/
 obj-$(CONFIG_IGBVF) += igbvf/
