--- a/drivers/i2c/busses/Kconfig	2024-06-01 01:38:16.554114492 +0200
+++ b/drivers/i2c/busses/Kconfig	2024-06-01 01:38:06.682082819 +0200
@@ -817,7 +817,7 @@

 config I2C_MT7621
 	tristate "MT7621/MT7628 I2C Controller"
-	depends on (RALINK && (SOC_MT7620 || SOC_MT7621)) || COMPILE_TEST
+	depends on (RALINK && (SOC_MT7620 || SOC_MT7621)) || ARCH_AIROHA || COMPILE_TEST
 	help
 	  Say Y here to include support for I2C controller in the
 	  MediaTek MT7621/MT7628 SoCs.
